File: CS/WindowsApplication3/RepositoryItemHtmlLabel.cs
using System;
using System.Collections.Generic;
using System.Windows.Forms;
using DevExpress.Skins;
using DevExpress.XtraEditors.Repository;
using DevExpress.XtraEditors.Registrator;
using System.ComponentModel;
using DevExpress.XtraEditors;
using DevExpress.XtraEditors.ViewInfo;
using DevExpress.Utils.Text;
using DevExpress.XtraEditors.Drawing;
using System.Drawing;
using DevExpress.Utils.Drawing;
using DevExpress.Utils;

namespace DXSample
{
    [UserRepositoryItem("RegisterCustomEdit")]
    public class RepositoryItemHtmlLabel : RepositoryItem
    {

        static RepositoryItemHtmlLabel() { RegisterCustomEdit(); }

        public RepositoryItemHtmlLabel() { }

        public const string CustomEditName = "CustomEdit";

        public override string EditorTypeName { get { return CustomEditName; } }

        public static void RegisterCustomEdit()
        {
            Image img = null;
            EditorRegistrationInfo.Default.Editors.Add(new EditorClassInfo(CustomEditName,
              typeof(HtmlLabel), typeof(RepositoryItemHtmlLabel),
              typeof(HtmlLabelViewInfo), new HtmlLabelPainter(), true, img));
        }

        private bool allowHtmlString;

        [DefaultValue(false)]
        public bool AllowHtmlString
        {
            get { return allowHtmlString; }
            set
            {
                if (allowHtmlString != value)
                {
                    allowHtmlString = value;
                    OnPropertiesChanged();
                }
            }
        }

        public override void Assign(RepositoryItem item)
        {
            BeginUpdate();
            try
            {
                base.Assign(item);
                RepositoryItemHtmlLabel source = item as RepositoryItemHtmlLabel;
                if (source == null) return;
                allowHtmlString = source.AllowHtmlString;
            }
            finally
            {
                EndUpdate();
            }
        }

        public static StringInfo CalculateStringInfo(Graphics g, AppearanceObject appearance, string text, Rectangle bounds ){
            return StringPainter.Default.Calculate(g, appearance, text, bounds);
        }
    }

    public class HtmlLabel : BaseEdit
    {

        static HtmlLabel() { RepositoryItemHtmlLabel.RegisterCustomEdit(); }

        public HtmlLabel() { }

        public override string EditorTypeName
        {
            get
            {
                return
RepositoryItemHtmlLabel.CustomEditName;
            }
        }

        [DesignerSerializationVisibility(DesignerSerializationVisibility.Content)]
        public new RepositoryItemHtmlLabel Properties
        {
            get { return base.Properties as RepositoryItemHtmlLabel; }
        }

    }
    public class HtmlLabelViewInfo : BaseEditViewInfo, IHeightAdaptable
    {
        public HtmlLabelViewInfo(RepositoryItem item) : base(item) { }

        #region IHeightAdaptable Members

        public int CalcHeight(GraphicsCache cache, int width)
        {
            RepositoryItemHtmlLabel item = Item as RepositoryItemHtmlLabel;
            if (item.AllowHtmlString)
            {
                StringInfo stringInfo = RepositoryItemHtmlLabel.CalculateStringInfo(cache.Graphics, PaintAppearance, DisplayText, ContentRect);
                return stringInfo.Bounds.Height;
            }
            CalcTextSize(cache.Graphics, true);
            return TextSize.Height;
        }

        #endregion

        public override AppearanceObject PaintAppearance
        {
            get
            {
                return base.PaintAppearance;
            }
            set
            {
                base.PaintAppearance = value;
                PaintAppearance.Options.UseTextOptions = true;
                PaintAppearance.TextOptions.WordWrap = DevExpress.Utils.WordWrap.Wrap;
            }
        }
    }

    public class HtmlLabelPainter : BaseEditPainter
    {
        public HtmlLabelPainter() : base() { }
        protected override void DrawContent(ControlGraphicsInfoArgs info)
        {
            Rectangle rect = new Rectangle(info.ViewInfo.ContentRect.X, info.ViewInfo.ContentRect.Y, info.ViewInfo.ContentRect.Width,
                info.ViewInfo.ContentRect.Height);
            info.Graphics.FillRectangle(info.ViewInfo.PaintAppearance.GetBackBrush(info.Cache), rect);
            HtmlLabelViewInfo viewInfo = info.ViewInfo as HtmlLabelViewInfo;
            RepositoryItemHtmlLabel item = viewInfo.Item as RepositoryItemHtmlLabel;
            if (item.AllowHtmlString)
            {
                StringInfo stringInfo = RepositoryItemHtmlLabel.CalculateStringInfo(info.Graphics, 
                    viewInfo.PaintAppearance, viewInfo.DisplayText, rect);
                StringPainter.Default.DrawString(info.Cache, stringInfo);
            }
            else
                info.Cache.DrawString(viewInfo.DisplayText, viewInfo.PaintAppearance.Font, viewInfo.PaintAppearance.GetForeBrush(info.Cache),
                    viewInfo.ContentRect, viewInfo.PaintAppearance.GetStringFormat());
        }
    }
}

File: VB/CustomHtmlTextEditor/WindowsApplication3/RepositoryItemHtmlLabel.vb
Imports DevExpress.XtraEditors.Repository
Imports DevExpress.XtraEditors.Registrator
Imports System.ComponentModel
Imports DevExpress.XtraEditors
Imports DevExpress.XtraEditors.ViewInfo
Imports DevExpress.Utils.Text
Imports DevExpress.XtraEditors.Drawing
Imports System.Drawing
Imports DevExpress.Utils.Drawing
Imports DevExpress.Utils

Namespace DXSample

    <UserRepositoryItem("RegisterCustomEdit")>
    Public Class RepositoryItemHtmlLabel
        Inherits RepositoryItem

        Shared Sub New()
            Call RegisterCustomEdit()
        End Sub

        Public Sub New()
        End Sub

        Public Const CustomEditName As String = "CustomEdit"

        Public Overrides ReadOnly Property EditorTypeName As String
            Get
                Return CustomEditName
            End Get
        End Property

        Public Shared Sub RegisterCustomEdit()
            Dim img As Image = Nothing
            Call EditorRegistrationInfo.Default.Editors.Add(New EditorClassInfo(CustomEditName, GetType(HtmlLabel), GetType(RepositoryItemHtmlLabel), GetType(HtmlLabelViewInfo), New HtmlLabelPainter(), True, img))
        End Sub

        Private allowHtmlStringField As Boolean

        <DefaultValue(False)>
        Public Property AllowHtmlString As Boolean
            Get
                Return allowHtmlStringField
            End Get

            Set(ByVal value As Boolean)
                If allowHtmlStringField <> value Then
                    allowHtmlStringField = value
                    OnPropertiesChanged()
                End If
            End Set
        End Property

        Public Overrides Sub Assign(ByVal item As RepositoryItem)
            BeginUpdate()
            Try
                MyBase.Assign(item)
                Dim source As RepositoryItemHtmlLabel = TryCast(item, RepositoryItemHtmlLabel)
                If source Is Nothing Then Return
                allowHtmlStringField = source.AllowHtmlString
            Finally
                EndUpdate()
            End Try
        End Sub

        Public Shared Function CalculateStringInfo(ByVal g As Graphics, ByVal appearance As AppearanceObject, ByVal text As String, ByVal bounds As Rectangle) As StringInfo
            Return StringPainter.Default.Calculate(g, appearance, text, bounds)
        End Function
    End Class

    Public Class HtmlLabel
        Inherits BaseEdit

        Shared Sub New()
            RepositoryItemHtmlLabel.RegisterCustomEdit()
        End Sub

        Public Sub New()
        End Sub

        Public Overrides ReadOnly Property EditorTypeName As String
            Get
                Return RepositoryItemHtmlLabel.CustomEditName
            End Get
        End Property

        <DesignerSerializationVisibility(DesignerSerializationVisibility.Content)>
        Public Overloads ReadOnly Property Properties As RepositoryItemHtmlLabel
            Get
                Return TryCast(MyBase.Properties, RepositoryItemHtmlLabel)
            End Get
        End Property
    End Class

    Public Class HtmlLabelViewInfo
        Inherits BaseEditViewInfo
        Implements IHeightAdaptable

        Public Sub New(ByVal item As RepositoryItem)
            MyBase.New(item)
        End Sub

#Region "IHeightAdaptable Members"
        Public Function CalcHeight(ByVal cache As GraphicsCache, ByVal width As Integer) As Integer Implements IHeightAdaptable.CalcHeight
            Dim item As RepositoryItemHtmlLabel = TryCast(Me.Item, RepositoryItemHtmlLabel)
            If item.AllowHtmlString Then
                Dim stringInfo As StringInfo = RepositoryItemHtmlLabel.CalculateStringInfo(cache.Graphics, PaintAppearance, DisplayText, ContentRect)
                Return stringInfo.Bounds.Height
            End If

            CalcTextSize(cache.Graphics, True)
            Return TextSize.Height
        End Function

#End Region
        Public Overrides Property PaintAppearance As AppearanceObject
            Get
                Return MyBase.PaintAppearance
            End Get

            Set(ByVal value As AppearanceObject)
                MyBase.PaintAppearance = value
                PaintAppearance.Options.UseTextOptions = True
                PaintAppearance.TextOptions.WordWrap = WordWrap.Wrap
            End Set
        End Property
    End Class

    Public Class HtmlLabelPainter
        Inherits BaseEditPainter

        Public Sub New()
            MyBase.New()
        End Sub

        Protected Overrides Sub DrawContent(ByVal info As ControlGraphicsInfoArgs)
            Dim rect As Rectangle = New Rectangle(info.ViewInfo.ContentRect.X, info.ViewInfo.ContentRect.Y, info.ViewInfo.ContentRect.Width, info.ViewInfo.ContentRect.Height)
            info.Graphics.FillRectangle(info.ViewInfo.PaintAppearance.GetBackBrush(info.Cache), rect)
            Dim viewInfo As HtmlLabelViewInfo = TryCast(info.ViewInfo, HtmlLabelViewInfo)
            Dim item As RepositoryItemHtmlLabel = TryCast(viewInfo.Item, RepositoryItemHtmlLabel)
            If item.AllowHtmlString Then
                Dim stringInfo As StringInfo = RepositoryItemHtmlLabel.CalculateStringInfo(info.Graphics, viewInfo.PaintAppearance, viewInfo.DisplayText, rect)
                StringPainter.Default.DrawString(info.Cache, stringInfo)
            Else
                info.Cache.DrawString(viewInfo.DisplayText, viewInfo.PaintAppearance.Font, viewInfo.PaintAppearance.GetForeBrush(info.Cache), viewInfo.ContentRect, viewInfo.PaintAppearance.GetStringFormat())
            End If
        End Sub
    End Class
End Namespace
